src/CoreFn/Expr.purs

```diff
@@ -131,69 +52,103 @@ data Expr a
-    Var unit <$> readQualified Ident value
-  readExpr' label _ = fail $ ForeignError $ "Unknown expression: " <> label
-
-readExprJSON :: String -> F (Expr Unit)
-readExprJSON = parseJSON >=> readExpr
+  show (Literal a l) =
+    "(Literal " <>
+      intercalate " " [ show a, show l ] <>
+    ")"
+  show (Constructor a t c fs) =
+    "(Constructor " <>
+      intercalate " " [ show a, show t, show c, show fs] <>
+    ")"
+  show (Accessor a s e) =
+    "(Accessor " <>
+      intercalate " " [ show a, show s, show e ] <>
+    ")"
+  show (ObjectUpdate a e fs) =
+    "(ObjectUpdate " <>
+      intercalate " " [ show a, show e, show fs ] <>
+    ")"
+  show (Abs a i e) =
+    "(Abs " <>
+      intercalate " " [ show a, show i, show e ] <>
+    ")"
+  show (App a e1 e2) =
+    "(App " <>
+      intercalate " " [ show a, show e1, show e2 ] <>
+    ")"
+  show (Var a q) =
+    "(Var " <>
+      intercalate " " [ show a, show q ] <>
+    ")"
+  show (Case a es cs) =
+    "(Case " <>
+      intercalate " " [ show a, show es, show cs ] <>
+    ")"
+  show (Let a bs e) =
+    "(Let " <>
+      intercalate " " [ show a, show bs, show e ] <>
+    ")"
+
 
 -- |
 --  A let or module binding.
 --
-data Bind a = Bind (Array (Tuple (Tuple a Ident) (Expr a)))
+data Bind a
+  = NonRec (Bind' a)
+  | Rec (Array (Bind' a))
+
+type Bind' a = Tuple (Tuple a Ident) (Expr a)
 
 derive instance eqBind :: Eq a => Eq (Bind a)
+-- derive instance functorBind :: Functor Bind
 derive instance ordBind :: Ord a => Ord (Bind a)
 
 instance showBind :: Show a => Show (Bind a) where
-  show (Bind x) = "(Bind " <> show x <> ")"
-
-readBind :: Foreign -> F (Bind Unit)
-readBind x = do
-  pairs <- objectProps x
-  bindings <- traverse fromPair pairs
-  pure $ Bind bindings
-
-  where
-
-  fromPair
-    :: { key :: String, value :: Foreign }
-    -> F (Tuple (Tuple Unit Ident) (Expr Unit))
-  fromPair pair = do
-    expr <- readExpr pair.value
-    let ident = Ident pair.key
-    pure $ Tuple (Tuple unit ident) expr
-
-readBindJSON :: String -> F (Bind Unit)
-readBindJSON = parseJSON >=> readBind
+  show (NonRec b) = "(NonRec " <> show b <> ")"
+  show (Rec b) = "(Rec " <> show b <> ")"
+
+
+-- |
+-- A guard is just a boolean-valued expression that appears alongside a set of binders
+--
+type Guard a = Expr a
+
+
+-- |
+-- An alternative in a case statement
+--
+data CaseAlternative a = CaseAlternative
+  -- |
+  -- A collection of binders with which to match the inputs
+  (Array (Binder a))
+  -- |
+  -- The result expression or a collect of guarded expressions
+  (Either (Array (Tuple (Guard a) (Expr a))) (Expr a))
+
+derive instance eqCaseAlternative :: Eq a => Eq (CaseAlternative a)
+derive instance ordCaseAlternative :: Ord a => Ord (CaseAlternative a)
+
+-- instance functorCaseAlternative :: Functor CaseAlternative where
+--   map f (CaseAlternative cabs car) = CaseAlternative
+--     (map (map f) cabs)
+--     (either (Left <<< map (map f *** map f)) (Right <<< map f) car)
```

Comment:
I still need to figure out how to satisfy the Functor instances that this depends on.

Reply:
While that `either` call is concise, I find it difficult to parse. A `case` expressions might be more better if you plan on having less experienced contributors in the future.

Reply:
I don’t disagree but I tried to keep this port as close as possible to the Haskell version so that nothing was accidentally lost in translation.

https://github.com/purescript/purescript/blob/f999179792ae2c49b2ad3019959081540da0ad9f/src/Language/PureScript/CoreFn/Expr.hs#L89-L93

Reply:
Got it. That explains the style. Thanks for the explanation and working on the port. It'll make adding backends easier in the future.